nav2_costmap_2d/plugins/static_layer.cpp

```diff
@@ -277,17 +276,13 @@ StaticLayer::interpretValue(unsigned char value)
 void
 StaticLayer::incomingMap(const nav_msgs::msg::OccupancyGrid::SharedPtr new_map)
 {
-  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
   if (!map_received_) {
-    map_received_ = true;
-    processMap(*new_map);
-  }
-  if (update_in_progress_.load()) {
-    map_buffer_ = new_map;
-  } else {
     processMap(*new_map);
```

Comment:
This will probably never be a problem, but please note that part of the `processMap` method is not thread-safe if we don't lock the Costmap2D mutex here as well. There is also a second call to this mutex within `processMap`, which might be able to produce the same deadlock that we are fixing.
The best way to go would probably find a way to move `processMap` to the main thread completely.
Again, this will probably never be a problem, but just in case.

Reply:
Actually, I removed it once, but I got some errors in the test.

```
[133.395s] The following tests FAILED:
[133.395s] 	 15 - test_collision_checker (Timeout)
[133.395s] 	 17 - inflation_tests (Timeout)
```

These tests use `LayeredCostmap` and `StaticLayer`, but do not call `updateMap` and wait for the costmap to be ready. So they got a timeout.

So I needed to keep this call here and keep it mutex free because the mutex can cause a deadlock at the very beginning.

Do you want me to remove the call and fix the tests too?

Reply:
I don't think you can simply remove it. If we want to remove it from here, we will need to do some other changes. There might be several places where the main loop might attempt to make use of the first map when was received, but not yet processed. We need to ensure that all places where the processed map is used are aware of the possibility of the map not being yet initialised.

Then, in my opinion, it is good that that test fails, it should, because you are changing the behaviour of this plugin. However, the tests failing are not a reason to not do the full change to processing in the main loop. If we want to move that last call to `processMap`, then we should fix these tests and/or write new ones that fit the current implementation.

>So I needed to keep this call here and keep it mutex free because the mutex can cause a deadlock at the very beginning.

I understand that you are focused on fixing the deadlock, but that doesn't mean you can simply leave unprotected code behind.
There are three points I want to do here, which I'm not sure have been considered here yet:

First, in [line 225 of the static layer](https://github.com/ros-planning/navigation2/blob/1d9d3c6208debf13b8ee38145bc25eef65a6b662/nav2_costmap_2d/plugins/static_layer.cpp#L225) we are locking the same mutex. This means you might not have fixed the lock entirely, only you don't see it anymore because it only happens with subsequent calls to `incomingMap`.

Second, [all this code](https://github.com/ros-planning/navigation2/blob/1d9d3c6208debf13b8ee38145bc25eef65a6b662/nav2_costmap_2d/plugins/static_layer.cpp#L189-L220) is retrieving data and performing operations on the layered costmap without protection, which to me, sounds like a very bad idea.

Lastly, this one I'm not sure about, but I think should be considered: setting `map_received_` without protection might bring undesired behaviour when checking it from the main loop. For example, in the main thread, you could have just called `updateBounds` with `map_received_` set to false, followed by a `updateCosts` with `map_received_` set to false.

As a rule of thumb, never leave code shared between threads unprotected.

Reply:
I don't quite understand the issue. In steady state, `processMap` is only called in the `updateBounds` function, so we're really only talking about the very first map when the current costmap is not set to any particular valid size/content.

In the situation we're talking about with the first map, the `map_received_` boolean at the start of both Update functions will make them immediately return without trying to do anything (if false) - which is before any map is processed. 

So, if no map is processed before updateBounds its OK. If its processed before that point in the map callback, also OK. 

If no map is processed before updateCosts and updateBounds, OK. If its processed after updateBounds but before updateCosts, that's where it could potentially be not OK. I think my comment just below lays out a solution to that (maintain the state of the "updatedness" from updateBounds to also use in updateCosts.